pkg/pillar/cmd/nim/controllerdns.go

```diff
@@ -70,28 +81,191 @@ func (n *nim) queryControllerDNS() {
+	go func() {
+		wg.Wait()
+		close(wgChan)
+	}()
+
+	select {
+	case <-wgChan:
+		var responses []DNSResponse
+		if countDnsRequests == 0 {
+			// fallback in case no resolver is configured
+			ips, err := net.LookupIP(domain)
+			if err != nil {
+				n.Log.Warnf("fallback resolver failed: %+v", err)
+			}
+			for _, ip := range ips {
+				responses = append(responses, DNSResponse{
+					IP:  ip,
+					Ttl: uint32(maxTTLSec),
+				})
+			}
+		}
+		return responses
+	case ip := <-resolvedIPsChan:
+		close(quit)
+		return ip
+	}
+}
+
+func (n *nim) dnsInfo(port types.NetworkPortStatus) (netmonitor.DNSInfo, bool) {
+	ifIndex, exist, err := n.networkMonitor.GetInterfaceIndex(port.IfName)
+	if !exist {
+		return netmonitor.DNSInfo{}, false
+	}
+	if err != nil {
+		n.Log.Warnf("converting ifName to ifIndex failed: %+v", err)
+		return netmonitor.DNSInfo{}, false
+	}
+
+	dnsInfo, err := n.networkMonitor.GetInterfaceDNSInfo(ifIndex)
+	if err != nil {
+		n.Log.Warnf("get interface dns info for %s failed: %+v", port.IfName, err)
+		return netmonitor.DNSInfo{}, false
+	}
+	return dnsInfo, true
+}
+
 // periodical cache the controller DNS resolution into /etc/hosts file
 // it returns the cached ip string, and TTL setting from the server
 func (n *nim) controllerDNSCache(etchosts, controllerServer []byte, ipaddrCached string) (string, int) {
-	if len(etchosts) == 0 || len(controllerServer) == 0 {
-		return ipaddrCached, maxTTLSec
-	}
-
 	// Check to see if the server domain is already in the /etc/hosts as in eden,
 	// then skip this DNS queries
-	if ipaddrCached == "" {
-		hostsEntries := bytes.Split(etchosts, []byte("\n"))
-		for _, entry := range hostsEntries {
-			fields := bytes.Fields(entry)
-			if len(fields) == 2 {
-				if bytes.Compare(fields[1], controllerServer) == 0 {
-					n.Log.Tracef("server entry %s already in /etc/hosts, skip", controllerServer)
-					return ipaddrCached, maxTTLSec
-				}
-			}
-		}
+	isCached, ipAddrCached, ttlCached := n.checkCachedEntry(etchosts, controllerServer, ipaddrCached)
+	if isCached {
+		return ipAddrCached, ttlCached
+	}
+
+	err := os.Remove(tmpHostFileName)
+	if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		n.Log.Warnf("%s exists but removing failed: %+v", tmpHostFileName, err)
 	}
 
+	var newhosts []byte
+	var ttlSec int
+	var lookupIPaddr string
+
+	dnsResponses := n.resolveWithPorts(string(controllerServer))
+	ttlSec = int(dnsResponses[0].Ttl)
+	lookupIPaddr = dnsResponses[0].IP.String()
+	serverEntry := fmt.Sprintf("%s %s\n", lookupIPaddr, controllerServer)
+	newhosts = append(etchosts, []byte(serverEntry)...)
+
+	if len(dnsResponses) == 0 {
+		newhosts = append(newhosts, etchosts...)
+	}
+
+	if len(dnsResponses) > 0 && n.writeTmpHostsFile(newhosts) {
+		n.Log.Tracef("append controller IP %s to /etc/hosts", lookupIPaddr)
+		ipaddrCached = lookupIPaddr
+	} else {
+		ipaddrCached = ""
+	}
+
+	return ipaddrCached, ttlSec
+}
+
+func (n *nim) writeTmpHostsFile(newhosts []byte) bool {
```

Comment:
I would just call it `writeHostsFile` - temporary file is just an intermediate step of the function.